paperqa/agents/search.py

```diff
@@ -326,19 +333,24 @@ async def _add_document() -> None:
             )
             raise
 
-    @staticmethod
+    async def checkpoint(self) -> None:
+        """Commit all pending changes, merge threads, and reset."""
+        if self._writer:
+            self._writer.commit()
+            self._writer.wait_merging_threads()
+            self._writer = None
```

Comment:
I am hesitant about the `if not self._writer` case where `commit` does nothing, it feels like a footgun.

Can we do something like `if not self._writer: raise RuntimeError("Please do XYZ before calling commit.")`

Or better yet, can we redesign the control flows so this case does not exist?

There is a design philosophy: the software works, because it must work

Reply:
I think we actually don't need it at all.
Either we `save_index` instantly after adding a document or we use the `changed` attribute